### DISCUSSION TITLE
Namespace URIs

### DISCUSSION THREAD

**Discussion by iamdrscott** [niemopen/ntac-admin#65] [General]
Does anyone remember why the existing NIEM namespace URIs always end in a slash?  I remember reading a reason for this, but now I can't find that reason anywhere.  

It might have something to do with relative URI resolution:

| Base URI                | Relative URI | Resolved absolute URI      |
| ----------------------- | ------------ | -------------------------- |
| http://example.com/foo/ | bar          | http://example.com/foo/bar |
| http://example.com/foo  | bar          | http://example.com/bar     |

That looks persuasive, except we don't really do anything with relative URIs and namespace documents, so far as I know.  So maybe we don't care.

The common practice for RDF vocabularies is to end in a "#" character, not "/#"; for example:

	http://www.w3.org/1999/02/22-rdf-syntax-ns#
	http://www.w3.org/2000/01/rdf-schema#
	http://www.w3.org/2002/07/owl#

So unless we have a good reason for the trailing slash, perhaps we should have

	http://docs.oasis-open.org/niemopen/ns/model/niem-core/6.0#
	or
	https://docs.oasis-open.org/niemopen/ns/model/niem-core/6.0/#

I don't see HTTPS in the [OASIS naming guidelines](https://docs.oasis-open.org/specGuidelines/namingGuidelines/resourceNaming.html).  Current practice is all over the map.  Take a look at this [StackOverflow page](https://stackoverflow.com/questions/33563814/can-https-urls-be-semantic-cooluris), especially the comments by Vladimir Alexiev.

**Comment by cdmgtri** [comment]
Originally, NIEM namespaces didn't have the trailing slash.  I think they were added in 3.0.

I didn't hear the reason, but my guess is that the NIEM URIs resolved to folders, not files, so the trailing slash made sense in that case.

The three RDF examples you listed all resolve to individual files, so no trailing slash makes sense for those.

NIEM 5.0 reorganized and flattened the directory structure, so we use redirects now to get us from the uri to the actual schema or folder.

Re OASIS naming guidelines - those are being updated.  I've asked about http vs https and was told that they plan to recommend https in an updated version and we could go ahead and use that.

**Reply by TomCarlson-NTAC** [reply]
They were indeed added in 3.0.

**Reply by davaya** [reply]
(Is there a naming guidelines 3.0 issued after v1.6 in Feb 2020?)

Interestingly, the OASIS naming guidelines [revision history](https://docs.oasis-open.org/specGuidelines/ndr/namingDirectives.html#revisionHistory) says: 

> **Version 1.6** Replaced all usage of "http:" with "https:"

which is true for everything in 1.6 *except* [namespaces](https://docs.oasis-open.org/specGuidelines/ndr/namingDirectives.html#xml-namespaces):

> An XML [namespace name](http://www.w3.org/TR/REC-xml-names/#iri-use) identified by an HTTP scheme URI reference must conform to the pattern: **http://docs.oasis-open.org/[tc-shortname]/ns/xxxx**

http://oasis-open.org will of course not result in an insecure connection.  Nor will any other http:// URL in today's web, since web servers routinely redirect http to https.  Embedded content like images may sometimes be optimized for speed by not using https, but that's the exception, not the rule.

**Comment by cabralje** [comment]
The latest version of the [OASIS Naming Directives](http://docs.oasis-open.org/specGuidelines/ndr/namingDirectives.html#xml-namespaces), includes a section on XML Namespaces. including the following rules for the final portion "xxxx" in the URI:

"xxxx" is a short string identifying an XML namespace, which should incorporate a versioning subcomponent (e.g., a string like 201011 representing a date or v1.1 representing a Version number)

**"xxxx" may use any of the same [sixty-four characters as names for files and directories/folders](http://docs.oasis-open.org/specGuidelines/ndr/namingDirectives.html#nameCharacters), plus internal "/", provided that it terminates with the character "/", "#", or an alphanumeric character [A-Za-z0-9].**

**Comment by TomCarlson-NTAC** [comment]
I think Christina's guess about resolving to folders is correct, if only because I vaguely remember a conversation about it, where I was pretty pedantic about them ending in slashes because otherwise a web server has to ask the browser "Did you really mean a folder instead?" Which no one cared about, but we ended up with the slash anyway.

**Comment by davaya** [comment]
[XML Namespaces](https://www.w3.org/TR/xml-names/#concepts) is extraordinarily cagey:

> Definition: An **expanded name** is a pair consisting of a [namespace name](https://www.w3.org/TR/xml-names/#dt-NSName) and a [local name](https://www.w3.org/TR/xml-names/#dt-localname).

But despite being prescriptive about comparing namespace names:

> URI references identifying namespaces are compared when determining whether a name belongs to a given namespace, and whether two names belong to the same namespace. 
Definition: The two URIs are treated as strings, and they are identical if and only if the strings are identical, that is, if they are the same sequence of characters.

the spec is absolutely silent about comparing URI references identifying **expanded names**.  I'd guess that that information vacuum is what led to the convention of including the terminating separator character in the namespace, so that if

> `fn` is [bound](https://www.w3.org/TR/xmlschema11-1/#ns-bindings) to http://www.w3.org/2005/xpath-functions

and

> `node-name`(https://www.w3.org/TR/xpath-functions/#accessors) is the localname of a function

then the URI that references `fn:node-name` can be unambiguously determined. The following are equivalent namespace:local pairs but the strings are different.  That ambiguity shouldn't exist

* http://www.w3.org/2005/xpath-functions/node-name
* http://www.w3.org/2005/xpath-functions#node-name
* http://www.w3.org/2005/xpath-functions:node-name

**Reply by TomCarlson-NTAC** [reply]
I read somewhere once that XML Namespaces are solely about adding an additional level of organization to element names, and nothing more. I think we see that in the definition of an expanded name, which is a _pair_ of two things, not a concatenation. From that perspective, there isn't a role in XML Namespaces themselves for string comparisons of namespace and local name concatenations. So that might be why there's no guidance. It's simply out of scope, and guidance will need to come from elsewhere.

However, this is all from memory and may be completely wrong.

**Comment by iamdrscott** [comment]
BLUF:  I did more research, and now I believe best practice has NIEM namespace URIs ending in "/", not "#".

A NIEM model namespace:

1. Has exactly one namespace URI, which is shared with no other namespace
2. Has a preferred namespace prefix
3. Has a governing authority
4. Defines a set of model components (classes and properties, also known as types, elements, and attributes)
5. Has an authoritative definition (XML Schema document and CMF file – these are equivalent and convertable)

A NIEM model namespace doesn't have a whole lot to do with the XML Namespace specification, especially in the post-XML-only world of NIEM 6.  But a NIEM model namespace is a pretty good match to a RDF vocabulary.  So we should look at best practices for those.  Which brings us to the W3C's [HashVsSlash](https://www.w3.org/wiki/HashVsSlash) wiki page from November 2022:

> There's a long-standing issue in the RDF community about whether to use **slash URIs** like —
>
> ```
> http://purl.org/dc/elements/1.1/title
> ```
>
> — or **hash URIs** like —
>
> ```
> http://www.w3.org/1999/02/22-rdf-syntax-ns#type
> ```
>
> That is, should the character before the last word be a hash ("#") or a slash ("/")?
>
> Short answer: **slash URIs are generally preferred** because they give greater flexibility, but read on for a more detailed explanation and comparative advantages of each approach.

Webb chose hash URIs back in 2014 when he wrote [5.6.1. Resource IRIs for XML Schema components and information items](https://reference.niem.gov/niem/specification/naming-and-design-rules/3.0/niem-ndr-3.0.html#section_5.6.1) in NDR 3.0.  

> Certain components defined by NIEM schemas and instances have corresponding resource IRIs. Each IRI is taken from a qualified name, as follows:
>
> - If namespace name ends with #: concatenate(namespace name, local part)
> - Otherwise: concatenate(namespace name, #, local part)

Before that, NIEM model components did not have URIs.  After that, NIEM model components had hash URIs.  Probably a good choice at the time.  But it led to some weirdness with NIEM JSON.  As you will recall, we chose JSON-LD as the basis for NIEM JSON.  JSON-LD is a RDF syntax; therefore, NIEM JSON is RDF.  The NDR defines a RDF mapping for NIEM XML; therefore, NIEM XML is also RDF.  ***That has to be the exact same RDF***, or we go to Jail, do not pass GO or collect $200. So for NIEM 5 JSON we have to write `@context` pairs like this:

```
"@context": {
  "nc": "http://release.niem.gov/niem/niem-core/5.0/#"  
}
```

See that trailing # character?  That has to be there, or else the NIEM JSON and NIEM XML won't be the same RDF.  Which means that in NIEM 5, the URI value in a NIEM JSON `@context` is *not the namespace URI*.  I am just sure that will bite us on the ass at some awkward moment.  Which is why until now I've said that NIEM 6 namespace URIs must end in a #.

HOWEVER, we could switch to slash URIs by rewriting that part of the NDR for NIEM 6, like this:

> Certain components defined by NIEM schemas and instances have corresponding resource IRIs. Each IRI is taken from a qualified name, as follows:
>
> - If namespace name ends with /: concatenate(namespace name, local part)
> - Otherwise: concatenate(namespace name, /, local part)

And then our NIEM JSON `@context` pairs will look like this:

```
"@context": {
  "nc": "http://docs.oasis-open.org/niemopen/ns/niem-core/6.0/"
}
```

The only users who will notice are those who are currently working with NIEM JSON and concerned about matching the RDF from NIEM XML.  I don't think there are any such users.

Finally, here is a cool example of what can happen when you try to resolve slash URIs.  We could eventually work our way up to this.

* a QUDT component:  http://qudt.org/schema/qudt/CurrencyUnit
* the QUDT namespace: http://qudt.org/schema/qudt

Resolving a NIEM namespace URI might return different formats based on content negotiation.  You could ask for *text/html* and get something readable in a browser.  You could ask for *application/niem+xsd* or *application/niem+cmf* and get the authoritative namespace definition in the desired formalism.  Someday.

**Comment by davaya** [comment]
From the wiki:
> **Short answer:** slash URIs are generally preferred because they give greater flexibility, but read on for a more detailed explanation and comparative advantages of each approach.

The longer answer is:
> **Advantages of using slash**
The vocabulary "[can be modularised while preserving a single namespace](https://lists.w3.org/Archives/Public/semantic-web/2022Nov/0024.html)". With hash, one document, whose URI is the everything up to the hash, has to contain information on all the things whose URIs are the same before the hash, i.e., sharing the same namespace. For a small hand-written ontology or data file this is fine, but for a machine-generated system it could be too large.

During the meeting there was discussion about whether to allow modularization of a namespace, and the answer then (perhaps carried forward from Webb's ghost) was "hell no" - we want a single document to define a single namespace.

I prefer slash and the ability for modules to contribute to a namespace, but it is more complex and error prone.  Software and guidance will have to be clear about how to properly use slash namespaces, and when duplicates are detected, ask "are you sure?"

Just so everyone goes in with eyes wide open.

**Reply by iamdrscott** [reply]
A NIEM namespace is completely defined by a single XML schema document.  I think we want to keep this rule.  Namespace modularization sounds to me like one namespace extending another, in the same way that the domains extend the core.  So I can imagine a future NIEM model with

- http://docs.oasis-open.org/niemopen/ns/model/niem-core/7.0/
- http://docs.oasis-open.org/niemopen/ns/model/domains/militaryOperations/7.0/
- http://docs.oasis-open.org/niemopen/ns/model/domains/militaryOperations/Land/7.0/
- http://docs.oasis-open.org/niemopen/ns/model/domains/militaryOperations/Air/7.0/

I see four namespaces, each defined by a single XML schema document... if we still support XSD as a modeling formalism in NIEM 7 :-)

**Reply by davaya** [reply]
That agrees with how I read the Wiki rationale - the four namespaces each defined by a single document are subcomponents of and modularly contribute to the greatest common factor namespace http://docs.oasis-open.org/niemopen/ns/model/.  That interpretation wouldn't be possible if any of the intermediate paths ended in a hash, e.g., http://docs.oasis-open.org/niemopen/ns/model/domains/militaryOperations#

And the existence of  
http://docs.oasis-open.org/niemopen/ns/model/domains/militaryOperations/Land/7.0/ doesn't preclude future definition of  
http://docs.oasis-open.org/niemopen/ns/model/domains/militaryOperations/Land/7.0/combat/ and  
http://docs.oasis-open.org/niemopen/ns/model/domains/militaryOperations/Land/7.0/logistics/, (each defined by a single XML schema document), though a NIEM rule could prohibit creating such extensions.  That rule could be enforced by ending each of the four original namespaces with #.

**Comment by davaya** [comment]
> Certain components defined by NIEM schemas and instances have corresponding resource IRIs. Each IRI is taken from a qualified name, as follows:
If namespace name ends with /: concatenate(namespace name, local part)
Otherwise: concatenate(namespace name, /, local part)

I'm not sure I understand this. I would interpret it as:
* If a URI ends with a /, it is a namespace name with no local part (analogous to a filesystem directory or non-leaf node)
* If a URI does not end with a separator, it is a resource name consisting of namespace and local part (analogous to a filesystem file or leaf node)

meaning that in the case of `http://purl.org/dc/elements/1.1/title`, the namespace is `http://purl.org/dc/elements/1.1/` and the local part is `title`.  I wouldn't concatenate "/" and some other local part after title.  If title is intended to be part of the namespace, the namespace must end in '/': `http://purl.org/dc/elements/1.1/title/`.

(Part of the complexity of modular namespaces is that they can be hierarchical - `http://purl.org/dc/elements/` could be a namespace with namespaces below it: `http://purl.org/dc/elements/1.1/`, `http://purl.org/dc/elements/1.2/`, `http://purl.org/dc/elements/2.0/`.  But there cannot be both a namespace and a resource with the same name: if `http://purl.org/dc/elements/1.1/title/` is a namespace then there can never be a `title` local part in the `http://purl.org/dc/elements/1.1/` namespace, and vice versa.)

**Reply by iamdrscott** [reply]
The only rule for a NIEM namespace identifier is that it must be an absolute URI.  We don't have a rule like this:

> * If a URI does not end with a separator, it is a resource name consisting of namespace and local part (analogous to a filesystem file or leaf node)

Maybe we should have that rule, but we don't.  So we have to cope with a NIEM extension schema document like this:

```
<xs:schema targetNamespace="http://example.com/no/trailing/slash" ...>
  <xs:complexType name="MyType" ...
```

The namespace name is `http://example.com/no/trailing/slash`.
The component URI should be `http://example.com/no/trailing/slash/MyType`.
Hence the rule

> If namespace name ends with /: concatenate(namespace name, local part)
> Otherwise: concatenate(namespace name, /, local part)

**Comment by iamdrscott** [comment]
Given the forthcomimg OASIS guidance from Chet, I now think that NIEM 5 URIs like

http://release.niem.gov/niem/niem-core/5.0/
will look like this in NIEM 6
https://docs.open-oasis.org/niemopen/ns/model/niem-core/6.0/
I think we're done here.
